src/rules/restrict-dependency-ranges.ts

```diff
@@ -0,0 +1,293 @@
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+
+import * as ESTree from "estree";
+import semver from "semver";
+
+import { createRule } from "../createRule.js";
+import { isJSONStringLiteral } from "../utils/predicates.js";
+
+const DEPENDENCY_TYPES = [
+	"dependencies",
+	"devDependencies",
+	"optionalDependencies",
+	"peerDependencies",
+];
+
+interface Option {
+	forDependencyTypes?: string[];
+	forPackages?: string[];
+	forVersions?: string;
+	rangeType: RangeType | RangeType[];
+}
+
+type Options = [Option | Option[] | undefined];
+
+const RANGE_TYPES = ["caret", "pin", "tilde"];
+type RangeType = "caret" | "pin" | "tilde";
+
+const schemaOptions = {
+	additionalProperties: false,
+	properties: {
+		forDependencyTypes: {
+			items: {
+				enum: DEPENDENCY_TYPES,
+			},
+			type: "array",
+		},
+		forPackages: {
+			items: {
+				type: "string",
+			},
+			type: "array",
+		},
+		forVersions: {
+			type: "string",
+		},
+		rangeType: {
+			oneOf: [
+				{
+					enum: RANGE_TYPES,
+				},
+				{
+					items: {
+						enum: RANGE_TYPES,
+					},
+					type: "array",
+				},
+			],
+		},
+	},
+	required: ["rangeType"],
+	type: "object",
+};
+
+const SYMBOLS = {
+	caret: "^",
+	pin: "",
+	tilde: "~",
+} satisfies Record<RangeType, string>;
+
+/**
+ * Given the original version, update it to use the correct range type.
+ */
+const changeVersionRange = (version: string, rangeType: RangeType): string => {
+	// We need to handle workspace versions with only the range indicator,
+	// slightly differently
+	if (/^workspace:[~^*]$/.test(version)) {
+		switch (rangeType) {
+			case "caret":
+				return "workspace:^";
+			case "pin":
+				return "workspace:*";
+			case "tilde":
+			default:
+				return "workspace:~";
+		}
+	}
+
+	return version.replace(
+		/^(workspace:)?(\^|~|<=?|>=?)?/,
+		`$1${SYMBOLS[rangeType]}`,
+	);
+};
+
+/**
+ * Check if the version is in a form that this rule supports.
+ */
+const isVersionSupported = (version: string): boolean => {
+	if (version === "*" || /^workspace:[*^~]$/.test(version)) {
+		return true;
+	}
+
+	const rawVersion = version.replace(/^workspace:/, "");
+	return !!semver.valid(rawVersion) || !!semver.validRange(rawVersion);
+};
+
+const capitalize = (str: string): string => {
+	return str.charAt(0).toUpperCase() + str.slice(1);
+};
+
+export const rule = createRule<Options>({
+	create(context) {
+		// Bail early if no options were provided
+		if (!context.options[0]) {
+			return {};
+		}
+
+		const optionsArray = Array.isArray(context.options[0])
+			? [...context.options[0]]
+			: [context.options[0]];
+
+		// Reverse the array, so that subsequent options override previous ones
+		optionsArray.reverse();
```

Comment:
[Refactor] Later on, the code calls `new RegExp(packageNamePattern)` for every `forPackages` entry x for every package. If a config has more complex patterns and a lot of package deps then that's going to end up being a lot of regexp creation and `.test()`ing of those newly created regexes.

How about pre-processing options early to normalize them to the format we expect? This'll also let us avoid the `rangeTypes` normalization later on:

```suggestion
		// Reverse the array, so that subsequent options override previous ones
		const optionsProvided = Array.isArray(context.options[0])
			? [...context.options[0]].reverse()
			: [context.options[0]];

		const options = optionsProvided.map((option) => ({
			...option,
			forPackages: option.forPackages?.map(
				(pattern) => new RegExp(pattern),
			),
			rangeTypes: Array.isArray(option.rangeType)
				? option.rangeType
				: [option.rangeType],
		}));
```

Reply:
Updated